### DISCUSSION TITLE
[closed] How to use DMA in earl's core together with PWM

### DISCUSSION THREAD

**Discussion by petervflocke** [earlephilhower/arduino-pico#1050] [Q&A] [answered]
Hi Earl,
Dear all,
I do like your arduino-pico core and appreciate the possibility to use [pico sdk](https://arduino-pico.readthedocs.io/en/latest/sdk.html)!
Apparently it is not as straightforward as I've initially thought.

I'd like to use DMA for PWM control, based on a great idea from [BirchJD](https://github.com/BirchJD/PicoRecPlayAudio).
This ([link](https://github.com/petervflocke/PicoPWMplay)) is a pico sdk based repo which produces a nice sound. PoC works as desired.

I tried to port it to your Arduino core with this code ([link](https://github.com/petervflocke/Pico-PWM-Sound)) and I failed. A simple "copy&paste" does not work. It produces no sound in compare to the pure c-based PoC code.

The main code to be ported is just a few lines in the file [WavPwmAudio.cpp](https://github.com/petervflocke/Pico-PWM-Sound/blob/master/src/WavPwmAudio.cpp) - same for both versions: sdk and arduino code.

I think I have read that Arduino uses some PWM channels, so we are not free to use all of them. Before I will spend Xmas on reading, I would like to ask for your opinion and or support. Do you see a method to port the code? This allows to add something more than beeps to the arduino pico projects, without any (almost) additional hardware.

Cheers and thank you in advance!
Peter

**Comment by earlephilhower** [comment]
Unfortunately the links in your post don't seem to work.  All seem 404'd.

Anyway, this is something I was thinking about myself but never got around to doing.  The idea is to set up a high frequency PWM signal on a pin and then use the DMA engine, on a timed transfer to send samples to it.

There are no PWM units used in the core unless you do `analogWrite` so it's all at your disposal.

I would start with the I2S library and the AudioRingBuffer.  You'll need to config the DMA to use a timer, not a FIFO free signal.  DMA target will be the PWM value reg for the associated pin, not the PIO FIFO.  The PWM channel will need to be config'd at the beginning and the audio samples received from the app will probably need to be scaled to the max PWM count.

It shouldn't be rocket science, but it probably is a good afternoon's hacking. :)

**Reply by petervflocke** [reply]
Thank You for this prompt answer!

> Unfortunately the links in your post don't seem to work. All seem 404'd.

Ooops, my bad, the two repos were "private", now they are public & accessible. Please check again if you like to see the code.

> The idea is to set up a high frequency PWM signal on a pin and then use the DMA engine, on a timed transfer to send samples to it.

Exactly and it works like a charm in pure "C".

> There are no PWM units used in the core unless you do analogWrite so it's all at your disposal.

What about "tone" e.g. tone(BuzzerPin,NOTE_B5,500); This applies PWM, doesn't it?

> You'll need to config the DMA to use a timer, not a FIFO free signal. DMA target will be the PWM value reg for the associated pin, not the PIO FIFO. The PWM channel will need to be config'd at the beginning and the audio samples received from the app will probably need to be scaled to the max PWM count.

I did not completely understand this approach, let me read it few more times :(. 
However, what would be the reason that the c-code does not work if all PWM-Units are free and arduino core does not mess up with the setup?

Best
Peter.

**Reply by earlephilhower** [reply]
`tone` uses a PIO, not a PWM.  I think it was my first PIO program, in fact. :)

I'll see if I can check out your code tomorrow, but I would suggest first checking you can properly set a single PWM output and control it from your code.  You can look at the `wiring_analog` file for a self-contained example.  https://github.com/earlephilhower/arduino-pico/blob/d1842743241a6a1294d2b1b26743c6ec693ae400/cores/rp2040/wiring_analog.cpp#L77-L120

**Reply by petervflocke** [reply]
Thank you, I will check the wiring_analog code snippet and try to adapt it.
Nevertheless, (any only if you'll find time), please look at my code as well.
Many thanks!
Peter

**Reply by Andy2No** [reply]
> please look at my code as well. Many thanks! Peter

I may be missing something, but are you trying to compile this in the Arduino IDE?

If so, change your main .c file to .ino, and change main() to setup(), with no parameters/arguments.

Normally, you'd put the contents of while(true) in to a function called loop(), but that's not essential.

.ino is c++, not c, so there may have to be some other small changes.  You may have to use *extern "C"* to reference functions in your other modules that are in c (see https://arduino.stackexchange.com/questions/946/how-to-call-c-functions-from-arduino-sketch), or change those files to .cpp

**Reply by petervflocke** [reply]
Hi Andy2No,
> I may be missing something, but are you trying to compile this in the Arduino IDE?

No, not at all. I do this in VSC/Platformio with following [setup](https://github.com/petervflocke/Pico-PWM-Sound/blob/master/platformio.ini). I do not have any compilation nor linking errors.
This is the added value of using Earls's core in compared to "standard" Arduino one, I can include/use pico SDK functions.

The code runs, but there is not any PWM or better to say soundable output, in compare to almost same basic code in pure c.
I will check later today with osci if there is any output at all. Yesterday it was too late :)

I know that my "try and error" approach is "dirty" but I simply wanted to add some nice sounds to my other project instead of tone-beeps but not extending the hardware. The above described converting WAV to PWM produced good quality sound and it looked very promising.

**Reply by Andy2No** [reply]
I see.  I've never used Platformio, so I probably can't help much.  We can use all of the SDK functions in standard Arduino though, AFAIK.  Some of them may need the right header files included.  I've used the gpio_.. ones a bit.

I see it needs a button, connecting gpio 14 (GPIO_KEY_PLAY) to ground when you press it, to trigger the sound via an interrupt and callback function.  What happens if you just start GpioKey off with that value instead?

```
static unsigned char GpioKey = GPIO_KEY_PLAY;
```

Otherwise, there would be no sound, if the callback isn't working, for some reason.  That seems like an easy thing to eliminate as the problem, anyway.

https://github.com/petervflocke/PicoPWMplay/blob/master/PicoPlayAudio.c

```
static unsigned char GpioKey = false;


int main(int argc, char* argv[])
{
   Init();
   while (true)
   {
      sleep_ms(100);
      gpio_put(PICO_DEFAULT_LED_PIN, (GpioKey != false));

      switch (GpioKey)
      {
         case GPIO_KEY_PLAY:
            WavPwmPlayAudio(DING_WAV);
            while (WavPwmIsPlaying());
            GpioKey = false;
            break;
      }
   };
}
```

You might want to comment out  *GpioKey = false;* in the loop too, so it keeps making a sound for long enough to set up hearing it / capturing it on a scope:

```

      switch (GpioKey)
      {
         case GPIO_KEY_PLAY:
            WavPwmPlayAudio(DING_WAV);
            while (WavPwmIsPlaying());
            // GpioKey = false;
            break;
      }
```

**Reply by earlephilhower** [reply]
Didn't run anything, but while eyeballing the Arduino example this line stood out:
https://github.com/petervflocke/Pico-PWM-Sound/blob/8367eba2206da2aedc88ba4c6aa4ba9578a25c1d/src/WavPwmAudio.cpp#L41

You want the slice for the GPIO passed in, not GPIO 0 which is hardcoded in it...

**Reply by petervflocke** [reply]
Hi Andy2No
hi Earl,

Many thanks for looking into the code. I do appreciate your interest and feedback!

TLDR: It works now, typo in GPIO definition, shame on me.
@earlephilhower I updated the GPIO in the pwm_gpio_to_slice_num as well to keep it consistent but the problem was in the output pin number.

@Andy2No 
Long version:
Actually you have analysed "c-version", which does work and I want to migrate to the Arduino framework.
It is enough to trigger the WavPwmPlayAudio once and the "magic" of DMA will push the complete table of data to the PWM in the background, leaving processor free to handle whatever we want. This is a perfect way to run sound and music effects in the background.

But in this test case I do nothing and wait only in the loop `while (WavPwmIsPlaying());` until the full sound data will be sent to PWM gpio. Changing the GpioKey to false switches the LED off.

Some not relevant side information to clarify the code logic:
In the GPIO_CallBack an interrupt routine the pressed GPIO number is assigned to GpioKey. This is an input of a potential bigger switch statement to handle more buttons. In this example there is just one case and one button. Please do not mixed low state of the active button with "false" in the variable.

Back to the core problem:
The migrated code https://github.com/petervflocke/Pico-PWM-Sound/blob/master/src/main.cpp has a simplified button handling, which also works but the main [WavPwmPlayAudio](https://github.com/petervflocke/Pico-PWM-Sound/blob/8367eba2206da2aedc88ba4c6aa4ba9578a25c1d/src/WavPwmAudio.cpp#L63) function does not drive the PWM.

It even runs as long as expected (the LED lights as long as the data is being processed, potentially sending the data somewhere, but it does not drive the PWM.

And now I  know why it did not work. Shame on me, shame on me, shame ....
I had a typo in the definition of the PWN output pin.
Wrong:
```C
#define BuzzerPin 16
#define GPIO_AUDIO_OUT_LEFT  6
```
And the right version:
```C
#define BuzzerPin 16
#define GPIO_AUDIO_OUT_LEFT  16
```
I wanted to test if I can combine tone and this DMA/PWM ton one pin to have simple beeps and nice wavs on one speaker. Yes I can.

By the way if you want to rebuild the "circuit", I added a circuit diagram [here](https://github.com/petervflocke/Pico-PWM-Sound/blob/master/docs/speaker.png) this produce low volume output. For a productive version we need here a simple amplifier either LM386 or PAM8403.

Lessons learned:

- Next time if I want to recycle an existing code I'll write it from scratch.
- Earls's core is active and provides support!

Thanks guys.
Peter

**Reply by petervflocke** [reply]
@earlephilhower 
Sorry I did not like my answer I gave you, that I had updated the code to keep it consistent - it is a stupid statement.
Actually I still haven't found 100% proper answer. For now it is like that "[me looking at the code](https://miro.medium.com/max/810/1*7wcPoKrXNYBZYTvDcHxhhA.png)"

The GPIO 0 hardcoded in the `PwmSliceNum = pwm_gpio_to_slice_num(0);` does not play any role, since the GPIO number is reflected in this [line](https://github.com/petervflocke/Pico-PWM-Sound/blob/5b34697654ed3deca7ec145136056c997c492641/src/WavPwmAudio.cpp#L83). 
`dma_channel_configure(WavPwmDmaCh, &WavPwmDmaChConfig, (void*)(PWM_BASE + PWM_CH0_CC_OFFSET), &(WavPwmData[2]), (WavPwmData[0] + (65536 * WavPwmData[1])) / 2, false);`

This reflects the target GPIO output pin:
`PWM_**CH0**_CC_OFFSET` points to `GPIO16` (CH0)
Switching to GPIO6 results in:
`PWM_**CH3**_CC_OFFSET` points to `GPIO6` (CH3)
... and so on according to:

1. https://github.com/majbthrd/pico-debug/blob/master/rp2040/hardware/regs/pwm.h
2. and the visual mapping: https://cdn-shop.adafruit.com/970x728/4864-04.png

For anybody googling this out, if you want to change the PWM output pin you have to reflect this in the `PWM_**CH??**_CC_OFFSET` value.

I still need to read the SDK and understand how  DMA and hardware PWM for RP2040 works. But now I have at least working code!

Cheers
Peter

**Reply by Andy2No** [reply]
I think you'd like the lectures by Hunter Adams, on YouTube.   I haven't watched them all, but what I have watched has been very informative, and he backs them up with code examples on his own website, in most cases:

This one covers using DMA with ADC.  There's a chapter index in the description:
https://www.youtube.com/watch?v=TGjUHChO1kM

This one covers hardware PWM:
https://www.youtube.com/watch?v=-OoGz9A6o2Q

**Reply by petervflocke** [reply]
Hi Andy,

> I think you'd like the lectures by Hunter Adams, on YouTube

Indeed, the complete channel looks very promising. I like the content of the lectures and the way of explanation. Thank you for your recommendation!

Have a great day!
Peter

**Reply by petervflocke** [reply]
Just in case you would like to see, or listen to: https://youtube.com/shorts/isAoaTnkfzQ

**Reply by Andy2No** [reply]
Nice :+1:

**Comment by earlephilhower** [comment, marked as answer]
Check #1076, @petervflocke .  I spent a couple afternoons adapting the I2S to PWM and it works pretty nicely IMHO!

**Reply by petervflocke** [reply]
Hi Earle,
"Hey Macarena" works well.
I'm wondering why you decided to implement a buffer as a list instead of simple "array" of data.
As long as the data come from a stable source e.g. flash, instead of network stream or slow sd card we can avoid processor and memory overhead to play it. I can imagine a ring buffer for unstable or slow sources. 
What do you think?

**Reply by earlephilhower** [reply]
The idea was to make it interchangeable w/the I2S setup where most folks are computing things on the fly (i.e. [decoding MP3s](https://github.com/earlephilhower/ESP8266Audio) or [generating speech](https://github.com/earlephilhower/ESP8266SAM)), so any little linked list overhead or memcpy would be minimal.  Even for PWM, you need to touch every work anyway (since WAVs are 16b signed and PWM is <16b and unsigned) so you're not really saving anything.

I can see potentially a "give me a buffer" or "take this whole buffer" API, but it's not high on my prio list.  PRs always appreciated. :)

**Reply by petervflocke** [reply]
> The idea was to make it interchangeable w/the I2S setup where most folks are computing things on the fly (i.e. [decoding MP3s](https://github.com/earlephilhower/ESP8266Audio) or [generating speech](https://github.com/earlephilhower/ESP8266SAM)),

I see! That makes absolutely sense. Thanks for your clarification and the links, especially for the speech generator. I need something similar for my project, where I use pre-generated samples for the time being.

> Even for PWM, you need to touch every work anyway (since WAVs are 16b signed and PWM is <16b and unsigned) so you're not really saving anything.

For a "fixed" background music and sound fx I convert WAVs to PWM offline, so I can point to buffer, fire and forget.

> I can see potentially a "give me a buffer" or "take this whole buffer" API, but it's not high on my prio list. PRs always appreciated. :)
I was thinking to tinker your code and add something like, "take this whole buffer" and play it for me and do not bother my precious process, sweating with sprites anyway.

Have a great day!
Peter

**Reply by earlephilhower** [reply]
Neat!  You'll probably still want to use chained, smaller buffers, of course, to keep the latency down when your little sprite needs to go "beep" or "boop." 😆
